File: Assets/FarmTap/Temple/Scripts/State/impactState.ts
import { _decorator, Component, Node, Vec3 } from 'cc';
import { ICharacterState } from './ICharacterState';
import { CharactorController } from '../Controller/CharactorController';
const { ccclass, property } = _decorator;

@ccclass('ImpactState')
export class ImpactState implements ICharacterState {
    enter(controller: CharactorController): void {
        // Handle entering impact state
        //controller.playAnimation(controller.impact);
    }

    update(controller: CharactorController): void {
        const current = controller.node.worldPosition;
        const dir = new Vec3();
        Vec3.subtract(dir, controller.safePos, current);

        const dist = dir.length();
        if (dist < 0.5) {
            controller.node.setWorldPosition(controller.safePos);
            return;
        }
        
        dir.normalize();
        const moveAmount = controller.moveSpeed * controller.getDeltaTime();
        const moveVec = dir.multiplyScalar(moveAmount);
        controller.node.setWorldPosition(controller.node.worldPosition.clone().add(moveVec));
    }

    exit(controller: CharactorController): void {
        // Handle exiting impact state
        //controller.playAnimation(controller.idle1111111111);
    }
}

File: Assets/FarmTap/Scripts/State/ImpactState.cs
using UnityEngine;

public class ImpactState : ICharacterState
{
    private float impactDuration = 0.3f;
    private float impactTimer = 0f;

    public void Enter(CharactorController controller)
    {
        Debug.Log("Enter Impact State");
        impactTimer = 0f;
        Gamecontroller.Instance.PlayEffect(controller.posEffect);
        controller.dustEffect.gameObject.SetActive(true);
        controller.dustEffect.Play();
       // SoundManager.instance.PlaySound(2);

    }

    public void Exit(CharactorController controller)
    {
        Debug.Log("Exit Impact State");
    }

    public void Update(CharactorController controller)
    {
        impactTimer += controller.GetDeltaTime();

        // Move back to safe position during impact
        Vector3 current = controller.transform.position;
        Vector3 targetSafePos = controller.SafePos;

        targetSafePos.y = current.y; 


        Vector3 dir = targetSafePos - current;
        float dist = dir.magnitude;

        if (dist < 0.5f)
        {
            Vector3 finalPos = controller.SafePos;
            finalPos.y = current.y;
            controller.transform.position = finalPos;

            // Transition back to Idle after impact duration
            if (impactTimer >= impactDuration)
            {
           //     controller.ChangeState(CharacterStateType.Idle);
            }
            return;
        }

        dir.Normalize();
        float moveAmount = controller.GetDeltaTime() * dist * 10f;
        Vector3 moveVec = dir * moveAmount;

        moveVec.y = 0;
        controller.transform.position = controller.transform.position + moveVec;
    }
}